utils/atomic-data-publisher/src/service.rs

```diff
@@ -0,0 +1,249 @@
+use helium_crypto::Keypair;
+use metrics_exporter_prometheus::PrometheusBuilder;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::signal;
+use tracing::{debug, error, info, warn};
+
+use crate::config::{ServiceConfig, Settings};
+use crate::database::DatabaseClient;
+use crate::errors::AtomicDataError;
+use crate::health_service::HealthService;
+use crate::metrics::MetricsCollector;
+use crate::polling_service::PollingService;
+use crate::publisher::AtomicDataPublisher as Publisher;
+
+#[derive(Debug)]
+pub struct AtomicDataPublisher {
+  database: Arc<DatabaseClient>,
+  publisher: Arc<Publisher>,
+  metrics: Arc<MetricsCollector>,
+  config: Settings,
+  shutdown_signal: tokio::sync::watch::Receiver<bool>,
+  pub shutdown_sender: tokio::sync::watch::Sender<bool>,
+}
+
+impl AtomicDataPublisher {
+  async fn validate_tables(database: &DatabaseClient, tables: &[String]) -> Result<(), AtomicDataError> {
+    for table_name in tables {
+      if !database.table_exists(table_name).await? {
+        return Err(AtomicDataError::ConfigError(format!(
+          "Required table '{}' does not exist",
+          table_name
+        )));
+      }
+    }
+    Ok(())
+  }
+
+  async fn init_database(database: &DatabaseClient, service_config: &ServiceConfig) -> Result<(), AtomicDataError> {
+    database.create_state_table().await?;
+
+    if service_config.polling_jobs.is_empty() {
+      warn!("No polling jobs configured");
+    }
+    Ok(())
+  }
+
+  pub async fn new(config: Settings) -> Result<Self, AtomicDataError> {
+    info!("Initializing Atomic Data Publisher service");
+
+    let metrics = Arc::new(MetricsCollector::new()?);
+
+    let database = Arc::new(
+      DatabaseClient::new_with_metrics(
+        &config.database,
+        config.service.polling_jobs.clone(),
+        Some(metrics.clone())
+      ).await?,
+    );
+
+    Self::validate_tables(&database, &config.database.required_tables).await?;
+    Self::init_database(&database, &config.service).await?;
+    database.init_polling_state().await?;
+    database.cleanup_stale_jobs().await?;
+
+    let keypair_path = config.signing.keypair_path.clone();
+    let keypair_data = if std::path::Path::new(&keypair_path).exists() {
+      std::fs::read(&keypair_path)?
+    } else {
+      return Err(AtomicDataError::ConfigError(format!(
+        "Keypair file not found at {}. Please provide a valid keypair file.",
+        keypair_path
+      )));
+    };
+
+    let keypair = Keypair::try_from(&keypair_data[..])
+      .map_err(|e| AtomicDataError::ConfigError(format!("Failed to load keypair from file: {}", e)))?;
+
+    info!("Using keypair with public key: {}", keypair.public_key());
+
+    let publisher = Arc::new(Publisher::new(
+      config.service.polling_jobs.clone(),
+      keypair,
+      config.service.clone(),
+      config.ingestor.clone(),
+      metrics.clone(),
+    ).await?);
+
+    let (shutdown_sender, shutdown_signal) = tokio::sync::watch::channel(false);
```

Comment:
There's a crate called `triggered` that is a nicer interface for shutdown signals than a `watch::channel()`